akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/WorkPullingProducerController.scala

```diff
@@ -0,0 +1,568 @@
+/*
+ * Copyright (C) 2019-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal.delivery
+
+import java.util.UUID
+import java.util.concurrent.ThreadLocalRandom
+
+import scala.concurrent.duration._
+import scala.reflect.ClassTag
+import scala.util.Failure
+import scala.util.Success
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.Behavior
+import akka.actor.typed.receptionist.Receptionist
+import akka.actor.typed.receptionist.ServiceKey
+import akka.actor.typed.scaladsl.ActorContext
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.LoggerOps
+import akka.actor.typed.scaladsl.StashBuffer
+import akka.util.Timeout
+import akka.actor.typed.internal.delivery.DurableProducerQueue.ConfirmationQualifier
+
+object WorkPullingProducerController {
+
+  sealed trait InternalCommand
+
+  sealed trait Command[A] extends InternalCommand
+
+  final case class Start[A](producer: ActorRef[RequestNext[A]]) extends Command[A]
+
+  final case class RequestNext[A](sendNextTo: ActorRef[A], askNextTo: ActorRef[MessageWithConfirmation[A]])
```

Comment:
For work pulling can the producer have multiple outstanding request nexts or is it still 1:1?

Reply:
There it's 1:1, but when PC receives a message there might not be any worker with demand any more, since they might have been deregistered in the meanwhile. Such are stashed.

Reply:
Does that mean in the WorkPulling case there's only one outstanding piece of work as there isn't a RequestNext back to the producer until it is confirmed? If i had a 1:N work pulling I'd expect at least N outstanding bits of work at any given time

Reply:
You can have N*request-window outstanding work.

WorkPullingProducer will send RequestNext to the application producer as long as there is any demand from any worker. Each ProducerController for a worker will send RequestNext to the WorkPullingProducer as long as there are more space in the request window (the `upToSeqNr` in the `Request` sent from the ConsumerController).

Reply:
Ah sorry i misinterpreted it being 1:1